runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java

```diff
@@ -0,0 +1,403 @@
+   * and Timers API</a> once it is available.
+   */
+  private static class ProcessFn<
+          InputT, RestrictionT, OutputT, TrackerT extends RestrictionTracker<RestrictionT>>
+      extends OldDoFn<KeyedWorkItem<String, KV<InputT, RestrictionT>>, OutputT>
+      implements OldDoFn.RequiresWindowAccess {
+    /**
+     * The state cell containing a watermark hold for the output of this {@link DoFn}. The hold is
+     * acquired during the first {@link DoFn.ProcessElement} call for this element and restriction,
+     * and is released when the {@link DoFn.ProcessElement} call returns {@link
+     * DoFn.ProcessContinuation#stop}.
+     *
+     * <p>A hold is needed to avoid letting the output watermark immediately progress together with
+     * the input watermark when the first {@link DoFn.ProcessElement} call completes.
+     *
+     * <p>The hold is updated with the future output watermark reported by ProcessContinuation.
+     */
+    private static final StateTag<Object, WatermarkHoldState<GlobalWindow>> watermarkHoldTag =
+        StateTags.makeSystemTagInternal(
+            StateTags.<GlobalWindow>watermarkStateInternal(
+                "hold", OutputTimeFns.outputAtLatestInputTimestamp()));
+
+    /**
+     * The state cell containing a copy of the element. Written during the first {@link
+     * DoFn.ProcessElement} call and read during subsequent calls in response to timer firings, when
+     * the original element is no longer available.
+     */
+    private final StateTag<Object, ValueState<WindowedValue<InputT>>> elementTag;
+
+    /**
+     * The state cell containing a restriction representing the unprocessed part of work for this
+     * element.
+     */
+    private StateTag<Object, ValueState<RestrictionT>> restrictionTag;
+
+    private final DoFn<InputT, OutputT> fn;
+    private transient DoFnInvoker<InputT, OutputT> invoker;
+
+    ProcessFn(
+        DoFn<InputT, OutputT> fn,
+        Coder<InputT> elementCoder,
+        Coder<? extends BoundedWindow> windowCoder) {
+      this.fn = fn;
+      elementTag =
+          StateTags.value("element", WindowedValue.getFullCoder(elementCoder, windowCoder));
+    }
+
+    @Override
+    public void setup() throws Exception {
+      invoker = DoFnInvokers.INSTANCE.newByteBuddyInvoker(fn);
+      restrictionTag =
+          StateTags.value("restriction", invoker.<RestrictionT>invokeGetRestrictionCoder());
+    }
+
+    @Override
+    public void processElement(final ProcessContext c) {
+      final WindowingInternals<KeyedWorkItem<String, KV<InputT, RestrictionT>>, OutputT> internals =
+          c.windowingInternals();
+      StateInternals<?> stateInternals = internals.stateInternals();
+      TimerInternals timerInternals = internals.timerInternals();
+
+      ValueState<WindowedValue<InputT>> elementState =
+          stateInternals.state(StateNamespaces.global(), elementTag);
```

Comment:
I'm not sure it matters -- the namespace just means that we're getting key-only state tags rather than a state tag associated with the window.

That said, it does raise a question -- if there are multiple input elements on a single key, this would break down. But, I think that we've rekeyed so that every element is on a unique key. We should document that as a precondition, and possibly mention it on the StateTags.

We may want to (in the future) discuss adding a state namespace which is just "unique for this element" or something.